[Kommander] cert-manager workaround (cops-related) (#4402)
mesosphere/dcos-docs-site: pages/dkp/kommander/2.2/release-notes/2.2.0/index.md

    ntials -o go-template='{{.data.username|base64decode}}:{{.data.password|base64decode}}')@$((kubectl -n kommander get cm konvoyconfig-kubeaddons -o go-template='{{if ne .data.clusterHostname ""}}{{.data.clusterHostname}}{{"\n"}}{{end}}' &&
    kubectl -n kommander get ingress gitea -o jsonpath="{.status.loadBalancer.ingress[0]['ip','hostname']}") | head -1)/dkp/kommander/git/kommander/kommander
    ```

1.  Change into the directory containing the credentials manifest: `kommander/shared/kommander-git-repository`

1.  Update the manifest file with what is stored in the cluster:

    ```bash
    kubectl -n kommander-flux get secret kommander-git-credentials -o yaml > git-credentials.yaml
    ```

1.  Delete the following fields from the `git-credentials.yaml` file:

    - metadata.labels
    - metadata.resourceVersion
    - metadata.uid
    - type

1.  Replace the content of the field "data.ca" file with the output of this command:

    ```bash
    kubectl -n kommander get secret git-tls -o jsonpath='{.data.ca\.crt}'
    ```

> <!-- Can we get an example output -->

Author: @palexster Since we are referring to replacing information based on the output of this command, I think it would be halpful to have an example output, if possible.

    while IFS= read -r cert; do
        signer=\$(kubectl get certificate -n \$namespace \$cert -o go-template='{{.spec.issuerRef.name}}')
        secret_name=\$(kubectl get certificate -n \$namespace \$cert -o go-template='{{.spec.secretName}}')
        # Workaround for the dex secret collisions

        if [[ \$signer =  "kommander-ca" ]]; then
            kubectl patch secret \$secret_name -n \$namespace --type json --patch '[{ "op": "replace", "path": "/data/ca.crt", "value": "'\$(kubectl get secret -n cert-manager kommander-ca -o go-template='{{index .data "ca.crt"}}')'" }]'
            ./cmctl renew -n \$namespace \$cert
        fi

    done <<< "\$certificates"
done <<< "\$namespaces"
EOF

chmod +x renew-certs.sh
./renew-certs.sh
```

Your cert-manager will renew your certificates successfully after 60 days, but the pods will use the previous certificates until day 90, so **you will have to restart your cluster anytime between days 60 and 90** after Kommander has been installed on your cluster (which usually coincides with the date you created your cluster), or after the last time you restarted your cluster.

> #### Restart your pods

Reviewer: ```suggestion
### Restart your pods
```

New clusters use the "delete first" strategy by default. Existing clusters are switched to the "delete first" strategy whenever their machines are updated with `update controlplane` and `update nodepool`.

### Cert-manager expiration workaround

> `cert-manager` renews all certificates 60 days after Kommander has been installed on your cluster. Unfortunately, some applications or pods fail to receive the renewed certificate information, causing them to break upon expiration (90 days after Kommander was installed, which usually coincides with the date the cluster was created). As a result, your cluster stops running, and in some cases, you are unable to access the UI.

Author: ```suggestion
`cert-manager` renews all certificates 60 days after Kommander has been installed on your cluster. Unfortunately, some applications or pods fail to receive the renewed certificate information, causing them to break upon expiration (90 days after Kommander was installed, which normally coincides with the date you created the cluster). As a result, your cluster stops running, and in some cases, you are unable to access the UI.
```

D2iQ provides a workaround that forces the applications to reconcile and recognize the renewed certificate. This workaround extends the validity of the certificates to 10 years, fixes the certification reload issue, and restarts the affected pods once the new certificate is issued.

It applies to any environment setup (networked, air-gapped, on-prem, etc.) and fixes the issue regardless of your issuer type ([SelfSigned][selfsigned] for air-gapped environments, [ACME][acme], or your own certificate issuer configured separately for your institution).

> To **prevent your applications from breaking**, or to **get them up and running again**, and fix this issue permanently, run this command:

Author: ```suggestion
To **prevent your applications from breaking**, or to **get the nodes up and running again**, and fix this issue permanently, run this command:
```
